generator/template.go

```diff
@@ -195,6 +195,97 @@ function getNotifyEntityArrivalSink<T>(notifyCallback: NotifyStreamEntityArrival
     }
   })
 }
+
+type RequestPayload = Record<string, unknown>;
+type Primitive = string | boolean | number;
+
+/**
+ * Checks if given value is of a primitive type
+ * @param  {unknown} value
+ * @return {boolean}
+ */
+function isPrimitive(value: unknown): boolean {
+  return (
+    typeof value === "string" ||
+    typeof value === "number" ||
+    typeof value === "boolean"
+  );
+}
+
+/**
+ * Flattens a deeply nested request payload and returns an object
+ * with only primitive values and non-empty array of primitive values
+ * @param  {RequestPayload} requestPayload
+ * @param  {String} path
+ * @return {Record<string, Primitive | Array<Primitive>}
+ */
+function flattenRequestPayload<T extends RequestPayload>(
+  requestPayload: T,
+  path: string = ""
+): Record<string, Primitive | Array<Primitive>> {
+  return Object.keys(requestPayload).reduce(
+    (acc: T, key: string): T => {
+      const value = requestPayload[key];
+      const newPath = path ? [path, key].join(".") : key;
+
+      const isObject =
+        Object.prototype.toString.call(value) === "[object Object]";
+      const isNonEmptyPrimitiveArray =
+        Array.isArray(value) &&
+        value.every(v => isPrimitive(v)) &&
+        value.length > 0;
+
+      switch (true) {
+        case isObject:
+          acc = {
+            ...acc,
+            ...flattenRequestPayload(value as RequestPayload, newPath)
+          };
+          break;
+        case isPrimitive(value) || isNonEmptyPrimitiveArray:
+          acc = { ...acc, [newPath]: value };
+          break;
+      }
+
+      return acc;
+    },
+    {} as T
+  ) as Record<string, Primitive | Array<Primitive>>;
+}
+
+/**
+ * Renders a deeply nested request payload into a string of URL search
+ * parameters by first flattening the request payload and then removing keys
+ * which are already present in the URL path.
+ * @param  {RequestPayload} requestPayload
+ * @param  {string[]} urlPathParams
+ * @return {string}
+ */
+export function renderURLSearchParams<T extends RequestPayload>(
+  requestPayload: T,
+  urlPathParams: string[] = []
+): string {
+  const flattenedRequestPayload = flattenRequestPayload(requestPayload);
+
+  const urlSearchParams = Object.keys(flattenedRequestPayload).reduce(
+    (acc: string[][], key: string): string[][] => {
+      // key should not be present in the url path as a parameter
+      const value = flattenedRequestPayload[key];
+      if (!urlPathParams.find(f => f === key)) {
+        if (Array.isArray(value)) {
+          acc = [...acc, ...value.map(m => [key, m.toString()])];
+        } else {
+          acc = [...acc, [key, value.toString()]];
+        }
+      }
+
+      return acc;
+    },
+    [] as string[][]
+  );
+
+  return "?" + new URLSearchParams(urlSearchParams).toString();
```

Comment:
there could be another case where in the url there's already a hardcoded part of query string specified, e.g. `test/{a}?predefined-query=a`. in that case  this is an `&` rather than a `?`.  I would suggest not worry about the `?` here and the leave it to the generator logic.

Reply:
In the final generated file, the renderURLSearchParams is part of the Path string and therefore I don't think there will be a case where there's already a hardcoded query string specified.
`/api/query/${req["a"]}${fm.renderURLSearchParams(req, ["a"])}`

`new URLSearchParams()` does not generate a `?` and the test fails if we do not prepend a `?` to the final generated URL Search Params.

```proto
  static HTTPGetWithURLSearchParams(req: HTTPGetWithURLSearchParamsRequest, initReq?: fm.InitReq): Promise<HTTPGetWithURLSearchParamsResponse> {
    return fm.fetchReq<HTTPGetWithURLSearchParamsRequest, HTTPGetWithURLSearchParamsResponse>(`/api/query/${req["a"]}${fm.renderURLSearchParams(req, ["a"])}`, {...initReq, method: "GET"})
  }
```

Reply:
give this a try in your test proto and you will see what I'm talking about.

```
option (google.api.http) = {
      get: "/api/query/{a}?test=3"
    };
```

Reply:
I see what you mean... then this means when we define the proto, we should add the `?` ourselves? Let me know if I understand your suggestion correctly

```
option (google.api.http) = {
      get: "/api/query/{a}?"
    };
```

Reply:
If we have to add the `?` ourselves then I think its a good idea to mention this in the README

Reply:
nah. I mean. in the generator code, we could tell if there's a query parameter has been specified before and it should be retained. and the query params that we add will need to adjust according to that.

say `get: "/api/query/{a}?test=3"` the query params we add will have `&` appended then with the params.
if just `get: "/api/query/{a}"` then the generator should add a `?` instead

Reply:
Aaah! I understand now. So we check this in the Go code and append `?` or `&` as required. Thank you for the detailed explanation 🙇🏼